pkg/image/prune/prune.go

```diff
@@ -526,38 +539,39 @@ func addBuildStrategyImageReferencesToGraph(g graph.Graph, strategy buildapi.Bui
 	case "ImageStreamImage":
 		_, id, err := imageapi.ParseImageStreamImageName(from.Name)
 		if err != nil {
-			glog.V(2).Infof("Error parsing ImageStreamImage name %q: %v - skipping", from.Name, err)
-			return
+			return fmt.Errorf("error parsing ImageStreamImage name %q: %v", from.Name, err)
 		}
 		imageID = id
 	case "DockerImage":
 		ref, err := imageapi.ParseDockerImageReference(from.Name)
 		if err != nil {
-			glog.V(2).Infof("Error parsing DockerImage name %q: %v - skipping", from.Name, err)
-			return
+			return fmt.Errorf("error parsing DockerImage name %q: %v", from.Name, err)
```

Comment:
I thought I saw agreement to make this a warning, not fail the entire prune process?  That would be my inclination.

Same for the ImageStreamImage check above.

If the reference is invalid, it clearly is not going to reference an image and keep us from pruning that image.

Reply:
> I thought I saw agreement to make this a warning, not fail the entire prune process? That would be my inclination.

@bparees Yes, but my answer was that you can't create such reference.

> If the reference is invalid, it clearly is not going to reference an image and keep us from pruning that image.

You can not be sure that this Strategy is not going to reference an image. You just don't know. I believe that in the conditions when we are not sure we need to save user data. In the end the user will see the error and can decide to remove this build/buildconfig or fix reference.

https://github.com/openshift/origin/pull/16717#discussion_r143468046

Reply:
> @bparees Yes, but my answer was that you can't create such reference.

I think relying on that guarantee is a mistake.  Bugs in the buildconfig validation could allow the creation of such a reference and then pruning is broken for no good reason.

> You can not be sure that this Strategy is not going to reference an image.

I don't care what it may do in the future.  I know it's not currently referencing an image.  What the user may update it to reference in the future when they fix their typo is irrelevant.  It may have tried to reference an image, but it is not referencing an image (it has an invalid reference, so it's not referencing anything).

> I believe that in the conditions when we are not sure we need to save user data.

What are we saving?  Can you give me a scenario where a user manages to create an image that we'd prune here despite an invalid reference existing in the system, that would ultimately be a problem?

If your argument is "the user meant to reference something but they did it wrong so we have to save everything until we know what they meant to reference", well what if they meant to reference "ruby" and they typed it as "riby"?  It's a "valid" reference, so it won't trip this error, but the "ruby" image is going to get cleaned up in that case too.  There's only so much protection we can offer.

> In the end the user will see the error and can decide to remove this build/buildconfig or fix reference.

The admin running this is not the same as the user who created the broken reference.  So your solution requires the admin to either go edit the user's object themselves, or track down the user and discuss how to resolve it.  Neither of those are ideal situations for an admin who's just trying to free up space (not to mention that this is frequently run as an automated job that may or may not have good monitoring in place).

In your referenced comment thread you said "done" but it's not obvious to me what that means.  You didn't add the suggested "force" flag and you didn't make this a warning.  (And to be clear, I prefer the latter).

Reply:
> I think relying on that guarantee is a mistake. Bugs in the buildconfig validation could allow the creation of such a reference and then pruning is broken for no good reason.

@bparees I'll change it, but it's your decision.

> In your referenced comment thread you said "done" but it's not obvious to me what that means

My `done` was about:

> If the error is printed, the admin will have the opportunity to act and tune/delete the object. Make sure to include namespace/name of the build config.

and I added the ability to see the name of the object.

Reply:
>  I'll change it, but it's your decision.

I asked a pretty direct question: 

What are we saving? Can you give me a scenario where a user manages to create an image that we'd prune here despite an invalid reference existing in the system, that would ultimately be a problem?

You seem very concerned about something here, but I haven't grasped what it is.  Can you help me understand the scenario you are envisioning where proceeding with pruning is a mistake and why it is a mistake?

I am trying to make an *educated* decision....  so far I am not seeing the risk in proceeding with pruning in this situation and i'd like you to help me be confident that I am not overlooking something.

Reply:
> I asked a pretty direct question:

@bparees At first I did not understand correctly. Sorry, of course I will try to explain my point of view.

> What are we saving?

Since this operation can not be canceled (we send requests to the registry to delete blobs/manifests), then we need to be very careful with the calculation of references. Otherwise, we will permanently delete objects that the user did not want to lose.

If we have references that we can not parse, it means that there is something wrong in the system (incomplete upgrade or something else). In this case, it is not the best time to perform actions that can not be rolled back.

Ignoring "broken" Build/BuildConfig will lead us to the image pruning if there are no other references in the cluster.

> Can you give me a scenario where a user manages to create an image that we'd prune here despite an invalid reference existing in the system, that would ultimately be a problem?

Imagine the situation when in the new version of origin we updated the `github.com/openshift/origin/pkg/image/reference` in such a way (it can be corner cases) that 10% of the old references stopped parsing. If you ignore the reference parsing error, you will lose 10% of the links in the graph. This will lead to a loss of actual data.

Reply:
> Imagine the situation when in the new version of origin we updated the github.com/openshift/origin/pkg/image/reference in such a way (it can be corner cases) that 10% of the old references stopped parsing. If you ignore the reference parsing error, you will lose 10% of the links in the graph. This will lead to a loss of actual data.

Thanks, that is a compelling argument.

Do we at least have enough information to add the namespace containing the bad buildconfig reference?  Right now i'm not convinced we're printing enough information to help an admin go find the bad reference and fix it.  They essentially would have to dump all the buildconfigs in all the namespaces and then grep for the string.  We should do more to help point them directly to where the problem is, in these cases.

Reply:
> Do we at least have enough information to add the namespace containing the bad buildconfig reference? 

@bparees Right now, I output an error to the user and specifying `<namespace>/<name>` of the Build/BuildConfig. Is this information enough ?

Reply:
yes, i hadn't seen that you were catching this error and wrapping it w/ additional information.  I see it now.